[WIP] Accessibles page modifications

knivets opened the pull request:
I copied the code from the `RoleGrantAcceptView` but I'm not sure why we need `existing_role` check. Considering that when the role grant is created the `user` is passed already https://github.com/djaodjin/djaodjin-saas/blob/master/saas/api/roles.py#L824.

smirolo commented:
The `existing_role` deals with the case where the user is already a manager for the organization, but now accept a contributor role. It used to be that a user could have multiple roles on an organization but as it turns out, it is best for managing complexity to constraint a user to have one and only one role at a time.

knivets commented:
I added a url to extras.py but the problem is that the view that deals with accessibles doesn't inherit from `OrganizationMixin` so I added the actual url to djaoapp and the one here mostly to keep testsite functioning. We need another solution for global urls.

smirolo commented on saas/api/roles.py:
```diff
@@ -944,3 +946,40 @@ def destroy(self, request, *args, **kwargs):
+        obj = get_object_or_404(get_role_model().objects.all(),
+                grant_key=serializer.validated_data['verification_key'])
+        existing_role = get_role_model().objects.filter(
+            organization=obj.organization, user=request.user).exclude(
```
Here we need to use `self.user` because the user is specified on the URL.

smirolo commented on saas/api/roles.py:
```diff
@@ -944,3 +946,40 @@ def destroy(self, request, *args, **kwargs):
+                    'organization': obj.organization.printable_name,
+                    'existing_role': existing_role.role_description.title})
+
+        obj.user = request.user       # We appropriate the Role here.
```
This cannot be happening here. That code is specific to a user clicking on a link in the invite e-mail and then login using a different account than the one invited.

smirolo commented on saas/urls/api/users.py:
```diff
@@ -28,13 +28,16 @@
 urlpatterns = [
     url(r'^legal/(?P<agreement>%s)/sign/$' % settings.ACCT_REGEX,
         AgreementSignAPIView.as_view(), name='saas_api_sign_agreement'),
+    url(r'^users/(?P<user>%s)/accessibles/accept/' % settings.MAYBE_EMAIL_REGEX,
```
The `verification_key` must be passed on the URL and not as a data parameter, otherwise we have an API which is not consistent with the other endpoints dealing with double opt-in. See
```    url(r'^profile/(?P<organization>%s)/subscribers/accept/'\
        '(?P<request_key>%s)/' % (ACCT_REGEX, VERIFICATION_KEY_RE),
        SubscriptionRequestAcceptAPIView.as_view(),
        name='saas_api_subscription_grant_accept'),
```

smirolo commented on saas/api/serializers.py:
```diff
@@ -597,15 +599,39 @@ class AccessibleSerializer(serializers.ModelSerializer):
 
+    def get_settings_url(self, obj):
+        org = obj.organization
+        abs_uri = self.context['request'].build_absolute_uri
```
Maybe though it seems incorrect if we look at the code in legal.py dealing with the same issue of getting an absolute uri.
```
        if settings.BUILD_ABSOLUTE_URI_CALLABLE:
            build_absolute_uri = import_string(
                settings.BUILD_ABSOLUTE_URI_CALLABLE)
            context.update({'site_url': build_absolute_uri(request)})
```

knivets replied:
So what should I do here?

smirolo commented on saas/static/js/djaodjin-saas-vue.js:
```diff
@@ -1954,6 +1954,16 @@ new Vue({
         url: djaodjinSettings.urls.user.api_accessibles,
         typeaheadUrl: djaodjinSettings.urls.api_candidates,
     },
+    methods: {
```
The code to accept a request for a subscription look like this (same file):
```
        acceptRequest: function(organization, request_key) {
            var vm = this;
            var url = (djaodjinSettings.urls.organization.api_profile_base +
                organization + "/subscribers/accept/" + request_key + "/");
            vm.reqPut(url, function (){
                vm.get();
            });
        },
```
In both cases `acceptGrant` or `acceptRequest`, the code should be identical except for the API end point being called.

knivets commented:
I basically copy pasted the code from the view that implement the logic, the one you told me about. As for the verification key not being included in the url - the request is not idempotent that’s why it made sense to use POST here.

knivets commented:
Also not sure what do you mean by 3rd point in the overall comments list.
